browser/static/js/canvas.js

```diff
@@ -67,9 +88,54 @@ class CanvasState {
     return this.scale * this.height;
   }
 
+  get paddedWidth() {
+    return this.scaledWidth + 2 * this.padding;
+  }
+
+  get paddedHeight() {
+    return this.scaledHeight + 2 * this.padding;
+  }
+
+  /**
+  * Creates 2D path for clipping to prevent other object methods (eg, the brush)
+  * from drawing in blank padding regions.
+  */
+  get visibleRegion() {
+    const region = new Path2D();
+    region.rect(this.padding, this.padding, this.scaledWidth, this.scaledHeight);
+    return region;
+  }
+
+  // only time trace needs to be accessed externally
+  // is to be passed to caliban.py in JSON format
+  get trace() {
+    return JSON.stringify(this._trace);
+  }
+
   // clear the current trace
   clearTrace() {
-    this.trace = [];
+    this._trace = [];
+  }
+
+  // addToTrace should not add duplicate coordinates
+  addToTrace() {
+    const newCoord = [this.imgY, this.imgX];
+    if (!this.inTrace(newCoord)) {
+      this._trace.push(newCoord);
+    }
+  }
+
+  // check if coordinate already exists in trace
+  inTrace(newCoord) {
+    let duplicate = false;
+    for (var i=0; i < this._trace.length; i++) {
+      let coord = this._trace[i];
+      if (coord[0] === newCoord[0] && coord[1] === newCoord[1]) {
+        duplicate = true;
+        break;
+      }
+    }
+    return duplicate;
```

Comment:
O(n) lookup speed on every mouse move seems inefficient. Let's brainstorm some other approaches. Could be the Coordinate object, which we can lookup in a map, or maybe we can set our own 2d boolean array and only add coords that are still False in the array?

Reply:
I'm not sure how to implement the Coordinate object + map, but I'm intrigued by the 2d boolean array idea